yt_idefix/data_structures.py

```diff
@@ -427,15 +427,217 @@ def _get_time(self):
+                    del uo_cache[unit]
+                else:
+                    ytLogger.info("Relying on %s: %s.", unit, uo_cache[unit])
+
+        bu = self.__BaseUnits(uo_cache)
+        for unit, value in bu.base_units.items():
+            setattr(self, unit, value)
 
-        self.mass_unit = density_unit * self.length_unit ** 3
-        self.time_unit = self.length_unit / self.velocity_unit
-        self.magnetic_unit = np.sqrt(4.0 * np.pi * density_unit) * self.velocity_unit
+        self.velocity_unit = self.length_unit / self.time_unit
+        self.density_unit = self.mass_unit / self.length_unit ** 3
+        self.magnetic_unit = (
+            np.sqrt(4.0 * np.pi * self.density_unit) * self.velocity_unit
+        )
         self.magnetic_unit.convert_to_units("gauss")
+        self.temperature_unit = self.quan(1.0, "K")
+
+    invalid_unit_combinations = [
+        {"magnetic_unit", "velocity_unit", "density_unit"},
+        {"velocity_unit", "time_unit", "length_unit"},
+        {"density_unit", "length_unit", "mass_unit"},
+    ]
+
+    default_units = {
+        "length_unit": "cm",
+        "time_unit": "s",
+        "mass_unit": "g",
+        "velocity_unit": "cm/s",
+        "magnetic_unit": "gauss",
+        "temperature_unit": "K",
+        # this is the one difference with Dataset.default_units:
+        # we accept density_unit as a valid override
+        "density_unit": "g/cm**3",
+    }
+
+    @classmethod
+    def _validate_units_override_keys(cls, units_override):
+        """Check that units in units_override are able to derive three base units:
+        mass, length and time
+        """
+
+        # YT supports overriding other normalisations, this method ensures consistency
+        # between supplied 'units_override' items and principles in PLUTO.
+
+        # PLUTO's normalisations/units have 3 degrees of freedom. Therefore, any combinations
+        # are valid other than the three cases listed explicitly.
+
+        if "temperature_unit" in units_override:
+            raise ValueError(
+                "Temperature is not allowed in units_override, "
+                "since it's always in Kelvin in PLUTO"
+            )
+
+        # Three units are enough for deriving others, more will likely cause conflict
+        if len(units_override) > 3:
+            raise ValueError(
+                "More than 3 degrees of freedom were specified "
+                f"in units_override ({len(units_override)} given)"
+            )
+
+        # check if provided overrides are allowed
+        suo = set(units_override)
+        if suo in cls.invalid_unit_combinations:
+            raise ValueError(
+                f"Combination {suo} passed to units_override "
+                "cannot derive all units\n"
+                f"Choose any other combinations, except for:\n {cls.invalid_unit_combinations}"
+            )
+
+        super(cls, cls)._validate_units_override_keys(units_override)
+
+    class __BaseUnits:
```

Comment:
I'm surprised you wrote this as a _nested_ class. I'd much rather have this class be seperated, otherwise I don't see the point. Also, a double trailing underscore is not a convention I'm familiar with. I think a single one is enough to indicate the class is private, isn't it ?

Reply:
Oh when I started writing this class, I wanted to include deriving other units as well. So it shouldn't work for other code and I considered it as a nested class. Now it becomes more general but I forgot to seperate it.

Do you think I should leave this class in this file, or move it to `definitions.py` and import it here? I prefer the latter.

Reply:
It doesn't need to be usable anywhere else to be separated. Otherwise you can leave it in the same file or move it to definitions if you think that's cleaner

Reply:
Well I decide to leave this work to the future when necessary.